File: HomeWork/fourthhomework/src/main/java/com/example/fourthhomework/DataAdapter.java
package com.example.fourthhomework;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import android.widget.Filter;
import android.widget.Filterable;

public class DataAdapter extends RecyclerView.Adapter<DataAdapter.ViewHolder> implements Filterable {
    private final List<Item> items;
    private final List<Item> itemsAll;
    private OnItemClickListener listener;

    interface OnItemClickListener {
        void onItemClick(Item item, int position);
    }

    public DataAdapter (List<Item> items, OnItemClickListener listener) {
        this.items = new ArrayList<>(items);
        this.itemsAll = new ArrayList<>(items);
        this.listener = listener;
    }

    public void setItems(List<Item> itemsNew) {
        items.clear();
        itemsAll.clear();
        items.addAll(itemsNew);
        itemsAll.addAll(itemsNew);
        notifyDataSetChanged();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        private TextView nameView, contactView;
        private ImageView imageView;
        ViewHolder(@NonNull View itemView) {
            super(itemView);
            nameView = itemView.findViewById(R.id.name);
            contactView = itemView.findViewById(R.id.contact);
            imageView = itemView.findViewById(R.id.image);
        }
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        Item item = items.get(position);
        holder.nameView.setText(item.getName());
        holder.contactView.setText(item.getContact());
        if (item.getImage().equals("email")) {
            holder.imageView.setImageResource(R.drawable.email);
        } else {
            holder.imageView.setImageResource(R.drawable.phone);
        }

        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override public void onClick(View v) {
                listener.onItemClick(item, position);
            }
        });
    }

    @Override
    public int getItemCount() {
        return items != null ? items.size() : 0;
    }

    @Override
    public Filter getFilter() {
        return myFilter;
    }

    private Filter myFilter = new Filter() {
        @Override
        protected FilterResults performFiltering(CharSequence charSequence) {
            List<Item> filteredList = new ArrayList<>();
            if (charSequence == null || charSequence.length() == 0) {
                filteredList.addAll(itemsAll);
            } else {
                for (Item item: itemsAll) {
                    if (item.getName().toLowerCase().contains(charSequence.toString().toLowerCase())) {
                        filteredList.add(item);
                    }
                }
            }
            FilterResults filterResults = new FilterResults();
            filterResults.values = filteredList;
            return filterResults;
        }

        @Override
        protected void publishResults(CharSequence charSequence, FilterResults filterResults) {
            items.clear();
            items.addAll((Collection<? extends Item>) filterResults.values);
            notifyDataSetChanged();
        }
    };
}

File: HomeWork/homework5_2/src/main/java/com/example/homework5_2/DataAdapter.kt
package com.example.homework5_2

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Filter
import android.widget.Filterable
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView

class DataAdapter(items: MutableList<Item>, listener: OnItemClickListener) : RecyclerView.Adapter<DataAdapter.ViewHolder>(), Filterable {
    private val items: MutableList<Item>
    private var itemsAll: MutableList<Item>
    private val listener: OnItemClickListener

    init {
        this.items = ArrayList(items)
        this.itemsAll = ArrayList(items)
        this.listener = listener
    }

    fun setItems(itemsNew: List<Item>) {
        items.clear()
        itemsAll.clear()
        items.addAll(itemsNew)
        itemsAll.addAll(itemsNew)
        notifyDataSetChanged()
    }

    interface OnItemClickListener {
        fun onItemClick(item: Item, position: Int)
    }

    class ViewHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val nameView: TextView = itemView.findViewById(R.id.name)
        val contactView: TextView  = itemView.findViewById(R.id.contact)
        val imageView: ImageView = itemView.findViewById(R.id.image)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val item: Item = items[position]
        holder.nameView.text = item.name
        holder.contactView.text = item.contact
        if (item.image == "email") {
            holder.imageView.setImageResource(R.drawable.email)
        } else {
            holder.imageView.setImageResource(R.drawable.phone)
        }
        holder.itemView.setOnClickListener { listener.onItemClick(item, position) }
    }

    override fun getItemCount() = items.size

    override fun getFilter() = myFilter

    private var myFilter: Filter = object : Filter() {
        override fun performFiltering(charSequence: CharSequence): FilterResults {
            var filteredList: MutableList<Item> = mutableListOf()
            if (charSequence.isEmpty()) {
                filteredList.addAll(itemsAll)
            } else {
                filteredList = itemsAll.filter{ item -> item.name.toLowerCase().contains(charSequence.toString().toLowerCase())}.toMutableList()
            }
            val filterResults = FilterResults()
            filterResults.values = filteredList
            return filterResults
        }
        override fun publishResults(charSequence: CharSequence, filterResults: FilterResults) {
            items.clear()
            items.addAll((filterResults.values as Collection<Item>))
            notifyDataSetChanged()
        }
    }
}